### PR TITLE
repro: persisting outputs invalidates build cache

### LINKED ISSUE #1852: run: make sure `persist` is ignoring build cache
We should not use build cache if there are any persistent outputs in the stage. Don't forget to put a warning on about it. See `Stage.is_cached`. Also need to not forget to reflect that in cmd help and our docs.

### PR DESCRIPTION
**Pull request by mroutis**
Fix #1852

### CODE REVIEW

File: dvc/repo/reproduce.py [hunk 1]
```diff
@@ -106,6 +106,10 @@ def _reproduce(
     stages = nx.get_node_attributes(G, "stage")
     node = os.path.relpath(stage.path, self.root_dir)
 
+    if any(out.persist for out in stage.outs):
+        logger.warning("Build cache is ignored when persisting outputs.")
+        force = True
```
Comment (efiop) [inline dvc/repo/reproduce.py]: Why are we forcing everything though?

File: tests/test_repro.py [hunk 2]
```diff
@@ -1404,3 +1405,40 @@ def test(self):
...
+class TestPersistentOutput(TestDvc):
+    def test_ignore_build_cache(self):
+        with open("immutable", "w") as fobj:
+            fobj.write("123456789")
+
+        warning = "Build cache is ignored when persisting outputs."
+
+        with self._caplog.at_level(logging.WARNING, logger="dvc"):
+            assert (
+                main(
```
Comment (efiop) [inline tests/test_repro.py]: This is hard to read, just use an intermediate `ret = main()` and then `assert ret == 0`.

File: tests/test_repro.py [hunk 3]
```diff
@@ -1404,3 +1405,37 @@ def test(self):
 
         expected_metrics_display = "{}: {}".format(metrics_file, metrics_value)
         self.assertIn(expected_metrics_display, self._caplog.text)
+
+
+class TestPersistentOutput(TestDvc):
```
Comment (efiop) [inline tests/test_repro.py]: This is not testing running `dvc run` twice, which is where `persist` should make second run ignore build cache and actually run the command.

**Comment by efiop** [comment]
For the record: Discussed this privately. Need to undo the `repro` part (notice discussion at https://github.com/iterative/dvc/issues/1884) and add a test for `run` part.

File: tests/func/test_run.py [hunk 4]
```diff
@@ -948,3 +949,35 @@ def test(self):
...
+            "--deps",
+            "immutable",
+            "--outs-persist",
+            "greetings",
+            "echo hello>>greetings",
+        ]
+
+        with self._caplog.at_level(logging.WARNING, logger="dvc"):
+            assert main(cmd) == 0
+            assert warning in self._caplog.text
```
Comment (efiop) [inline tests/func/test_run.py]: There shouldn't be a warning during the first run.

Comment (mroutis) [reply tests/func/test_run.py]: @efiop , should we skip the _overwrite stage file_ confirmation when `outs-persist` is used?

File: tests/func/test_run.py [hunk 5]
```diff
@@ -948,3 +949,31 @@ def test(self):
...
+            "--overwrite-dvcfile",
+            "--deps",
+            "immutable",
+            "--outs-persist",
+            "greetings",
+            "echo hello>>greetings",
+        ]
+
+        with self._caplog.at_level(logging.WARNING, logger="dvc"):
+            assert main(cmd) == 0
```
Comment (efiop) [inline tests/func/test_run.py]: should we test that that warning is not printed during the first run now? :slightly_smiling_face:

Comment (mroutis) [reply tests/func/test_run.py]: sure :+1:

File: tests/func/test_run.py (hunk 4 above)
Comment (efiop) [reply tests/func/test_run.py]: @mroutis why would we want that though? :slightly_smiling_face:

File: tests/func/test_run.py (hunk 5 above)
Comment (mroutis) [reply tests/func/test_run.py]: done, @efiop  (sorry for such back-and-forth, I should've test it correctly beforehand :sweat_smile:)

**Pull request merged**